api/tests/iq_views_tests.py

```diff
@@ -0,0 +1,178 @@
+import hashlib
+import phonenumbers
+import pytest
+import responses
+from unittest.mock import Mock, patch
+
+from twilio.rest import Client
+
+from django.conf import settings
+
+from rest_framework.test import RequestsClient
+
+from phones.tests.models_tests import make_phone_test_user
+from api.tests.phones_views_tests import _make_real_phone, _make_relay_number
+
+
+API_ROOT = "http://127.0.0.1:8000"
+INBOUND_SMS_PATH = f"{API_ROOT}/api/v1/inbound_sms_iq/"
+
+
+pytestmark = pytest.mark.skipif(not settings.IQ_ENABLED, reason="IQ_ENABLED is False")
+
+
+@pytest.fixture()
+def phone_user(db):
+    yield make_phone_test_user()
+
+
+@pytest.fixture(autouse=True)
+def mocked_twilio_client():
+    """
+    Mock PhonesConfig with a mock twilio client
+    """
+    with patch(
+        "phones.apps.PhonesConfig.twilio_client", spec_set=Client
+    ) as mock_twilio_client:
+        mock_fetch = Mock(
+            return_value=Mock(
+                country_code="US", phone_number="+12223334444", carrier="verizon"
+            )
+        )
+        mock_twilio_client.lookups.v1.phone_numbers = Mock(
+            return_value=Mock(fetch=mock_fetch)
+        )
+        yield mock_twilio_client
+
+
+def test_iq_endpoint_missing_verificationtoken_header():
+    client = RequestsClient()
+    response = client.post(INBOUND_SMS_PATH)
+    assert response.status_code == 401
+    response_body = response.json()
+    assert "missing Verificationtoken header" in response_body["detail"]
+
+
+def test_iq_endpoint_missing_messageid_header():
+    client = RequestsClient()
+    client.headers.update({"Verificationtoken": "valid"})
+    response = client.post(INBOUND_SMS_PATH)
+
+    assert response.status_code == 401
+    response_body = response.json()
+    assert "missing MessageId header" in response_body["detail"]
+
+
+def test_iq_endpoint_invalid_hash():
+    message_id = "9a09df23-01f3-4e0f-adbc-2a783878a574"
+    client = RequestsClient()
+    client.headers.update({"Verificationtoken": "invalid value"})
+    client.headers.update({"MessageId": message_id})
+    response = client.post(INBOUND_SMS_PATH)
+
+    assert response.status_code == 401
+    response_body = response.json()
+    assert "verficiationToken != computed sha256" in response_body["detail"]
+
+
+def test_iq_endpoint_valid_hash_no_auth_failed_status():
+    message_id = "9a09df23-01f3-4e0f-adbc-2a783878a574"
+    combined = settings.IQ_INBOUND_API_KEY + message_id
+    token = hashlib.sha256(combined.encode()).hexdigest()
```

Comment:
Should we extract this to a function so that the test and the code can share the auth logic? And make it so that just the function changes if the IQ HMAC calc is fixed?

Reply:
moved into a `compute_iq_mac` function.